File: src/main/java/com/github/svegon/capi/event/render/UnderwaterVisibilityCallback.java
package com.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.network.ClientPlayerEntity;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

@FunctionalInterface
public interface UnderwaterVisibilityCallback {
    Event<UnderwaterVisibilityCallback> EVENT = EventFactory.createArrayBacked(UnderwaterVisibilityCallback.class,
            (player, callback) -> {}, listeners -> (player, callback) -> {
        for (UnderwaterVisibilityCallback listener : listeners) {
            listener.onUnderwaterVisibility(player, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
    });

    void onUnderwaterVisibility(ClientPlayerEntity player, CallbackInfoReturnable<Float> callback);
}

File: src/client/kotlin/io/github/svegon/mclientapi/client/event/render/UnderwaterVisibilityCallback.kt
package io.github.svegon.mclientapi.client.event.render

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.network.ClientPlayerEntity
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable

fun interface UnderwaterVisibilityCallback {
    fun onUnderwaterVisibility(player: ClientPlayerEntity, callback: CallbackInfoReturnable<Float>)

    companion object {
        @JvmField
        val EVENT: Event<UnderwaterVisibilityCallback> = EventFactory.createArrayBacked<UnderwaterVisibilityCallback>(
            UnderwaterVisibilityCallback::class.java,
            UnderwaterVisibilityCallback { player: ClientPlayerEntity, callback: CallbackInfoReturnable<Float> -> }) { listeners: Array<UnderwaterVisibilityCallback> ->
            UnderwaterVisibilityCallback { player: ClientPlayerEntity, callback: CallbackInfoReturnable<Float> ->
                for (listener in listeners) {
                    listener.onUnderwaterVisibility(player, callback)

                    if (callback.isCancelled) {
                        return@UnderwaterVisibilityCallback
                    }
                }
            }
        }
    }
}
